keps/sig-cli/kubectl-commands-in-headers.md

```diff
@@ -0,0 +1,219 @@
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Requests sent to the apiserver from kubectl include http request headers with context about the kubectl command that
+created the request.  This information could be used by cluster administrators for debugging or
+to gather telemetry about how users are interacting with the cluster.
+
+
+## Motivation
+
+Kubectl generates requests sent to the apiserver for commands such as `apply`, `delete`, `edit`, `run`, however
+the context of the command for the requests is lost and unavailable to cluster administrators.  Context would be
+useful to cluster admins both for debugging the cause of requests as well as providing telemetry about how users
+are interacting with the cluster.
+
+### Goals
+
+- Allow cluster administrators to identify how requests in the logs were generated from
+  kubectl commands.
+
+### Non-Goals
+
+NA
+
+## Proposal
+
+Include in requests made from kubectl to the apiserver:
+
+- the kubectl subcommand
+- which flags were specified (but not the values)
+- enum values for stdin and stdout
+- a generated session id
+
+### Kubectl-Command Header
+
+The `Kubectl-Command` Header contains the kubectl sub command.  It contains
+the path to the subcommand (e.g. `kubectl apply`) to disambiguate sub commands
+that might have the same name and different paths.
+
+Examples:
+
+- `Kubectl-Command: kubectl apply`
+- `Kubectl-Command: kubectl create secret tls` 
+- `Kubectl-Command: kubectl delete`
+- `Kubectl-Command: kubectl get`
+
+### Kubectl-Flags Header
+
+The `Kubectl-Flags` Header contains a list of the kubectl flags that were provided to the sub
+command.  It does *not* contain the flag values, only the names of the flags.  It
+does not normalize into short or long form.  If a flag is provided multiple times
+it will appear multiple times in the list.  Flags are sorted alpha-numerically and
+separated by a ','.
+
+Examples:
+
+- `Kubectl-Flags: --filename,--recursive`
+- `Kubectl-Flags: -f,-f,-f,-R` 
+- `Kubectl-Flags: --dry-run,-o`
+
+### Kubectl-Input Header
+
+The `Kubectl-Input` Header contains the types of input used.  Because the flag values are not
+provided, this can be used to determine if the command input was from STDIN, Local Files or
+Remote Files.  Format is a list of the types of input provided.
+
+Examples:
+
+- `Kubectl-Input: stdin`
+- `Kubectl-Input: file` 
+- `Kubectl-Input: http`
+- `Kubectl-Input: file,stdin`
+- `Kubectl-Input: file,file,http`
+
+### Kubectl-Output Header
+
+The `Kubectl-Output` Header contains the type of output used.  Because the flag values are
+not provided, this can be used to determine if the output is yaml,json,go-template,wide.
+Note: it does *not* contain non-enumeration values, such as the custom-columns values for
+for custom-columns output.
+
+Examples:
+
+- `Kubectl-Output: yaml`
```

Comment:
This will overlap with `Kubectl-flags` or you're planning to exclude output-related flags?

Reply:
Or this will let us differentiate between default output and requested output, like you're describing in your example?

Reply:
Right.  I could actually move these into flags as enums.

Reply:
Updated